packages/ckeditor5-engine/src/model/utils/insertcontent.ts

```diff
@@ -306,7 +358,7 @@ class Insertion {
 			return Range._createOn( this._nodeToSelect );
 		}
 
-		return this.model.schema.getNearestSelectionRange( this.position );
+		return this.model.schema.getNearestSelectionRange( this._affectedEnd ? this._affectedEnd.toPosition() : this.position );
```

Comment:
I wouldn't make the change here.

First, although it seems logical in most cases, I am not 100% sure that `this._affectedEnd` is always "same enough" as `this.position`. It seems it should be but it is not apparent from the code. Actually, the logic behind setting `this._affectedEnd` seems complicated. When `this._affectedEnd` is set, it is always set to `this.position` but then, `this.position` may change differently than `this._affectedEnd`.

Second, `getSelectionRange()` makes sense only straight after handling nodes insertion. Most probably it would even make more sense to return the selection range in `handleNodes()` rather than have a separate method. What happens in your code is this: you insert nodes, then you tinker with the inserted content outside of `Insertion` class and then you expect `Insertion` class to return correct results. It kind-of worked for you, because `this._affectedEnd` is a `LiveRange`, so it got updated after you removed the nodes.

But it is not a good design, as you make change in `Insertion` class that is triggered by a new functionality added in a different place. Unless it is a trivial change, an unavoidable change, or a bug we should solve it in a different place. It seems that results returned by `getSelectionRange()` make sense and there were no bugs related strictly to how `Insertion` class works.

Instead, you should reorganize the code, so that the new functionality (handling markers) solves the problem that it introduces. After all, the selection is wrong because we inserted and removed the fake elements:

1.  `handleNodes()`
2.  `let newRange = getSelectionRange()`
3.  if there are fake markers elements to handle:
    1.  create a new live range, same as the range returned by `getSelectionRange()`
    2.  remove fake elements, create markers
    3.  overwrite `newRange` with the live range (that will be updated in step 3.ii)